File: resources/xcode/NativePHP/NativeRender/NativeRootTabsRenderer.swift
import SwiftUI

/// Native chrome renderer for the `native_root_tabs` sentinel. Uses
/// SwiftUI's `TabView` — system tab styling, selection
/// animation, and (on iOS 26+) Liquid Glass material on the bar by
/// default.
///
/// **Tab content semantics.** TabView wants every tab view alive at
/// once, but PHP only ever has the active tab's tree in hand. So:
///   - Inactive tabs render `Color.clear` placeholders (zero memory)
///   - Active tab renders the screen content child (the non-tab,
///     non-action child of the sentinel)
///   - The "active" index is read from `bottom_nav_item.active` (the
///     framework's `TabBar::highlight()` already sets exactly one
///     tab's `active` based on the current URI)
///
/// **Selection sync.**
///   - User taps a tab → SwiftUI's `selection` updates → onChange fires
///     `sendPressEvent` on that tab's `onPress` (auto-wired by
///     `BottomNavItem::resolveProps()` to a `replace` navigation toward
///     the tab's URL) → PHP publishes the new tree at the new URL →
///     re-rendered tree has `active = true` on the just-tapped tab →
///     activeTabIdx updates → set `selection` to match (no-op, already
///     there).
///   - PHP-initiated tab swap (e.g. programmatic `replace()`): same
///     publish flow → activeTabIdx changes → set `selection` → SwiftUI
///     animates the swap. We don't fire press (already at activeTabIdx).
///
/// **Three-tier appearance** (matches NavigationStack):
///   - No `background_color` → defaults; on iOS 26+ Liquid Glass.
///   - `background_color` set → `.toolbarBackground(.visible, .tabBar)`
///     with the explicit color → opaque native bar.
///   - Inline `<native:bottom-nav>` blade bypasses entirely.
struct NativeRootTabsRenderer: View {
    let node: NativeUINode

    @State private var selection: Int = 0
    @StateObject private var tabBag = TabCoordinatorBag()

    var body: some View {
        let tabs = node.children.filter { $0.type == "bottom_nav_item" }
        let accessory = node.children.first { $0.type == "tab_accessory" }
        let currentUri = node.props.getString("current_uri", default: "")
        let minimizeOnScroll = node.props.getBool("minimize_on_scroll")

        // Determine which tab "owns" the current URI via longest URL
        // prefix match against tab URLs. This is what lets a pushed
        // sub-route (e.g. `/syncup-native/chat/123` under a
        // `/syncup-native` Chats tab) render INSIDE the Chats tab's
        // NavigationStack rather than full-screen-replacing the layout.
        // Falls back to `BottomNavItem.active` (TabBar::highlight set
        // the same way) so non-prefix tab layouts still pick correctly.
        let owningIdx = owningTabIndex(for: currentUri, tabs: tabs)
            ?? (tabs.firstIndex { $0.props.getBool("active") } ?? 0)

        // Sync the owning tab's coordinator with this publish — cache
        // synchronously so the destination resolver always sees the
        // freshest tree, then defer the path mutation (push / pop) via
        // async since `path` is @Published.
        if !currentUri.isEmpty, owningIdx < tabs.count {
            let owningRootUri = tabs[owningIdx].props.getString("url", default: "")
            let coord = tabBag.coordinator(forIdx: owningIdx, rootUri: owningRootUri)
            coord.cache(uri: currentUri, node: node)
            DispatchQueue.main.async {
                coord.receive(uri: currentUri, rootNode: node)
            }
        }

        let activeArgb = node.props.getColor("active_color", default: 0)
        let activeColor: Color = activeArgb != 0 ? Color(argb: activeArgb) : .accentColor
        let bgArgb = node.props.getColor("background_color", default: 0)
        let isDark = node.props.getBool("dark")

        // Folded NavBar config — present when the layout supplies both
        // bars. Tabs render an inner NavigationStack with toolbar even
        // when this is empty (so per-tab pushes still get a back chevron).
        let navBack = node.props.getBool("nav_back")
        let navTitleText = node.props.getString("nav_title", default: "")
        let navBgArgb = node.props.getColor("nav_background_color", default: 0)
        let navTextArgb = node.props.getColor("nav_text_color", default: 0)
        let hasNavBar = navBack || !navTitleText.isEmpty

        // Explicit `return` switches the body out of @ViewBuilder mode so
        // the side-effect `if !currentUri.isEmpty { … }` block above is a
        // plain statement (not a "view" with a `()` result) — same pattern
        // as `NativeRootStackRenderer.body`.
        return TabView(selection: $selection) {
            // Key the ForEach off the enumerated `offset` (integer
            // position) rather than the SwiftUI element id — element
            // ids regenerate on every PHP publish, which would make
            // ForEach treat every tab as brand-new on tab switch.
            // SwiftUI would then rebuild every Tab + its inner
            // NavigationStack from scratch each publish, producing a
            // visible flicker (toolbar styles applied a frame late)
            // and recreating animation state mid-transition. Tab
            // positions are stable across publishes for a layout's
            // tabBar(), so the offset is a reliable identity.
            ForEach(Array(tabs.enumerated()), id: \.offset) { idx, tab in
                let label = tab.props.getString("label", default: "")
                let icon = tab.props.getString("icon", default: "circle")
                let isSearch = tab.props.getBool("search")

                // `Tab(role: .search)` is the iOS 18+ API; on iOS 26 the
                // role triggers the floating Liquid Glass capsule beside
                // the main bar (Apple's Photos / Music pattern). Pre-26
                // it's a semantic no-op visually but kept for consistency.
                let tabRootUri = tab.props.getString("url", default: "")
                let coord = tabBag.coordinator(forIdx: idx, rootUri: tabRootUri)

                Tab(value: idx, role: isSearch ? .search : nil) {
                    PerTabContent(
                        coordinator: coord,
                        hasNavBar: hasNavBar,
                        fallbackTitle: navTitleText,
                        fallbackShowBack: navBack,
                        fallbackTextArgb: navTextArgb,
                        fallbackBgArgb: navBgArgb
                    )
                } label: {
                    Label(label, systemImage: getIconForName(icon))
                }
                .badge(badgeFor(tab))
                    // ↑ `.badge(Text?)` — nil means "no badge". Passing
                    //   an empty string here would render as a presence
                    //   dot on the new `Tab(value:role:)` API.
            }
        }
        .tint(activeColor)
        .modifier(ExplicitBarBackgroundModifier(
            argb: bgArgb,
            placement: .tabBar
        ))
        .preferredColorScheme(isDark ? .dark : nil)
        .modifier(TabBarLiquidGlassModifier(
            accessory: accessory,
            minimizeOnScroll: minimizeOnScroll
        ))
        .onAppear {
            selection = owningIdx
        }
        .onChange(of: owningIdx) { newIdx in
            // PHP-driven owning-tab change (publish landed under a
            // different tab's URL prefix) — sync our SwiftUI selection.
            if selection != newIdx {
                selection = newIdx
            }
        }
        .onChange(of: selection) { newIdx in
            // User tapped a tab — fire its press handler so PHP runs
            // the BottomNavItem-auto-wired `replace` navigation. We
            // don't fire when selection already matches the owning tab
            // (PHP-driven sync path).
            guard newIdx != owningIdx, newIdx < tabs.count else { return }
            let tab = tabs[newIdx]
            if tab.onPress != 0 {
                NativeElementBridge.sendPressEvent(tab.onPress, nodeId: tab.id)
            }
            // Action-only tab (no URL) — the press fires something
            // off-screen (e.g. a bottom sheet) instead of navigating,
            // so PHP won't republish to switch the owning tab. Snap
            // selection back to the actual owning tab so the visible
            // "selected" indicator doesn't get stuck.
            let url = tab.props.getString("url", default: "")
            if url.isEmpty {
                DispatchQueue.main.async {
                    selection = owningIdx
                }
            }
        }
    }

    /// Longest URL prefix match against tab URLs. A pushed sub-route
    /// (e.g. `/syncup-native/chat/123`) returns the index of the tab
    /// whose URL is the longest prefix of `currentUri`. Returns nil if
    /// no tab claims the URI; the caller should fall back to PHP's
    /// `BottomNavItem.active` flag.
    private func owningTabIndex(for currentUri: String, tabs: [NativeUINode]) -> Int? {
        guard !currentUri.isEmpty else { return nil }
        var bestIdx: Int? = nil
        var bestLen: Int = -1
        for (idx, tab) in tabs.enumerated() {
            let tabUrl = tab.props.getString("url", default: "")
            guard !tabUrl.isEmpty else { continue }
            let isMatch = (currentUri == tabUrl) || currentUri.hasPrefix(tabUrl + "/")
            if isMatch && tabUrl.count > bestLen {
                bestIdx = idx
                bestLen = tabUrl.count
            }
        }
        return bestIdx
    }

    /// On the new `Tab(value:role:)` API, `.badge` accepts a `Text?` — nil
    /// means "no badge". Passing an empty string would render an empty
    /// red presence bubble (the iOS 18+ behavior), so explicit-text vs.
    /// nil is the only correct distinction.
    ///   - `badge` prop set → show that string
    ///   - `news` flag set → show a bullet (renders as a red dot)
    ///   - neither → return nil so no badge is drawn at all
    private func badgeFor(_ tab: NativeUINode) -> Text? {
        let badge = tab.props.getString("badge", default: "")
        if !badge.isEmpty { return Text(badge) }
        if tab.props.getBool("news") { return Text("•") }
        return nil
    }
}

/// One tab's content area — hosts its own NavigationStack bound to the
/// per-tab coordinator's path. Each pushed level inside the tab reads
/// its toolbar config (title, back, actions, colors) from its own
/// cached node, so a chat detail (pushed) gets its own title + actions
/// independently of the tab root's toolbar.
///
/// `fallback*` props come from the layout's folded NavBar — used only
/// before the tab's coordinator has cached anything (cold-start frame
/// for an inactive tab) so the inner NavigationStack still renders
/// something with a sensible toolbar.
private struct PerTabContent: View {
    @ObservedObject var coordinator: PerTabNavigationCoordinator
    let hasNavBar: Bool
    let fallbackTitle: String
    let fallbackShowBack: Bool
    let fallbackTextArgb: Int
    let fallbackBgArgb: Int

    var body: some View {
        if hasNavBar {
            NavigationStack(path: $coordinator.path) {
                levelView(uri: coordinator.rootUri, isRoot: true)
                    .navigationDestination(for: String.self) { uri in
                        levelView(uri: uri, isRoot: false)
                    }
            }
            .onChange(of: coordinator.path) { newPath in
                coordinator.onPathChange(newPath: newPath)
            }
        } else {
            levelContent(for: coordinator.rootNodeCache[coordinator.rootUri])
        }
    }

    @ViewBuilder
    private func levelView(uri: String, isRoot: Bool) -> some View {
        if let cached = coordinator.rootNodeCache[uri] {
            renderLevel(cached, isRoot: isRoot)
        } else {
            Color.clear
                .navigationTitle(fallbackTitle)
                .navigationBarTitleDisplayMode(.inline)
                .modifier(TabsToolbarModifier(
                    showBack: fallbackShowBack && isRoot,
                    title: fallbackTitle,
                    actions: [],
                    textArgb: fallbackTextArgb,
                    bgArgb: fallbackBgArgb
                ))
        }
    }

    @ViewBuilder
    private func renderLevel(_ root: NativeUINode, isRoot: Bool) -> some View {
        let title = root.props.getString("nav_title", default: fallbackTitle)
        // Manual back chevron only shows at the tab's root level — at
        // that level there's no NavigationStack history to pop, so the
        // chevron fires `sendSystemBackEvent` to leave the tabs entirely
        // (back to the launcher / wherever the user came from).
        // Pushed levels get NavigationStack's automatic system chevron
        // which pops the path natively.
        let layoutShowBack = root.props.getBool("nav_back") || fallbackShowBack
        let manualBack = layoutShowBack && isRoot
        let textArgb = root.props.getColor("nav_text_color", default: fallbackTextArgb)
        let bgArgb = root.props.getColor("nav_background_color", default: fallbackBgArgb)
        let actions = root.children.filter { $0.type == "top_bar_action" }
        // Bottom-pinned content (chat input, search bar, etc.) — extracted
        // out of children so it doesn't render inline; the actual content
        // is the BottomBar wrapper's first child.
        let bottomBar = root.children.first { $0.type == "bottom_bar" }
        let screenContent = root.children.first {
            $0.type != "bottom_nav_item"
                && $0.type != "top_bar_action"
                && $0.type != "tab_accessory"
                && $0.type != "bottom_bar"
        }

        levelContent(for: root, screenContent: screenContent)
            .navigationTitle(title)
            .navigationBarTitleDisplayMode(.inline)
            .modifier(TabsToolbarModifier(
                showBack: manualBack,
                title: title,
                actions: actions,
                textArgb: textArgb,
                bgArgb: bgArgb
            ))
            // Hide the parent TabView's tab bar on any pushed level
            // FIRST — matches Apple's iMessage / Music / Mail pattern
            // where pushed levels (chat detail, now-playing) get the
            // full screen. Crucially, this changes the available safe
            // area, so any subsequent `safeAreaInset(.bottom)` (the
            // bottom-bar modifier below) pins to the correct edge.
            // Inverting the order leaves the bar latched to where the
            // tab bar *used to be* — visually mid-screen.
            .modifier(HideTabBarOnPushModifier(isRoot: isRoot))
            .modifier(BottomBarInsetModifier(bottomBar: bottomBar))
    }

    @ViewBuilder
    private func levelContent(for cached: NativeUINode?, screenContent: NativeUINode? = nil) -> some View {
        let content: NativeUINode? = screenContent ?? cached?.children.first {
            $0.type != "bottom_nav_item"
                && $0.type != "top_bar_action"
                && $0.type != "tab_accessory"
                && $0.type != "bottom_bar"
        }
        if let content {
            // GlassEffectContainer scopes :interactive press animations.
            // See NativeRootStackRenderer.screenView for the full rationale.
            NodeView(node: content).withGlassContainer()
        } else {
            Color.clear
        }
    }
}

/// The per-level toolbar (back chevron + actions) plus background /
/// color-scheme modifiers, factored out so the SwiftUI type-checker
/// doesn't time out chasing the chained modifiers in renderLevel.
private struct TabsToolbarModifier: ViewModifier {
    let showBack: Bool
    let title: String
    let actions: [NativeUINode]
    let textArgb: Int
    let bgArgb: Int

    func body(content: Content) -> some View {
        let textColor: Color = textArgb != 0 ? Color(argb: textArgb) : .primary
        let toolbarScheme: ColorScheme? = {
            guard textArgb != 0 else { return nil }
            let r = Double((textArgb >> 16) & 0xFF) / 255.0
            let g = Double((textArgb >>  8) & 0xFF) / 255.0
            let b = Double( textArgb        & 0xFF) / 255.0
            let luminance = 0.299 * r + 0.587 * g + 0.114 * b
            return luminance > 0.5 ? .dark : .light
        }()

        content
            .toolbar {
                ToolbarItem(id: "back", placement: .topBarLeading) {
                    if showBack {
                        Button {
                            NativeElementBridge.sendSystemBackEvent()
                        } label: {
                            Image(systemName: "chevron.backward")
                                .font(.system(size: 17, weight: .semibold))
                                .foregroundColor(textColor)
                        }
                    }
                }
                ToolbarItemGroup(placement: .topBarTrailing) {
                    ForEach(actions) { action in
                        TabsActionView(action: action, textColor: textColor)
                    }
                }
            }
            .toolbarColorScheme(toolbarScheme, for: .navigationBar)
            .modifier(ExplicitBarBackgroundModifier(
                argb: bgArgb,
                placement: .navigationBar
            ))
    }
}

/// Single trailing toolbar action — plain Button when no sub-items,
/// pull-down Menu when `NavAction::items()` produced sub-actions.
private struct TabsActionView: View {
    let action: NativeUINode
    let textColor: Color

    var body: some View {
        let icon = action.props.getString("icon", default: "ellipsis")
        let subItems = action.children.filter { $0.type == "top_bar_action" }

        if subItems.isEmpty {
            Button {
                if action.onPress != 0 {
                    NativeElementBridge.sendPressEvent(action.onPress, nodeId: action.id)
                }
            } label: {
                Image(systemName: getIconForName(icon))
                    .font(.system(size: 17, weight: .semibold))
                    .foregroundColor(textColor)
            }
        } else {
            Menu {
                ForEach(subItems) { item in
                    let itemLabel = item.props.getString("label", default: "")
                    let itemIcon = item.props.getString("icon", default: "")
                    let isDestructive = item.props.getBool("destructive")
                    Button(role: isDestructive ? .destructive : nil) {
                        if item.onPress != 0 {
                            NativeElementBridge.sendPressEvent(item.onPress, nodeId: item.id)
                        }
                    } label: {
                        if !itemIcon.isEmpty {
                            Label(itemLabel, systemImage: getIconForName(itemIcon))
                        } else {
                            Text(itemLabel)
                        }
                    }
                }
            } label: {
                Image(systemName: getIconForName(icon))
                    .font(.system(size: 17, weight: .semibold))
                    .foregroundColor(textColor)
            }
        }
    }
}

/// Pins a `bottom_bar` element above the safe-area-bottom (and above
/// the keyboard when one is presented). Renders nothing if the layout
/// didn't supply a bottom bar for this level.
///
/// Uses `.safeAreaInset(.bottom)` — the battle-tested primitive that
/// reserves space above the safe-area-bottom and renders the bar inside
/// it. Keyboard avoidance is automatic. Bar content is responsible for
/// its own visible chrome (bg color or glass class).
private struct BottomBarInsetModifier: ViewModifier {
    let bottomBar: NativeUINode?

    func body(content: Content) -> some View {
        // Always use `.safeAreaInset(.bottom)` — the battle-tested
        // primitive for pinning content above the safe-area bottom.
        //
        // Earlier this path branched to `.safeAreaBar(.bottom)` on iOS 26+
        // for the floating-glass treatment iMessage uses. That implementation
        // mis-pins on the iOS 26 simulator (bar latches mid-screen) and adds
        // a wide glass plate behind the bar that visually washes content
        // above it. Both symptoms cleared by reverting to `.safeAreaInset`.
        // Re-enable `.safeAreaBar` later if Apple's implementation
        // stabilises on real hardware AND we want the floating treatment.
        if let bottomBar, let inner = bottomBar.children.first {
            content.safeAreaInset(edge: .bottom, spacing: 0) {
                NodeView(node: inner)
            }
        } else {
            content
        }
    }
}

/// Hides the enclosing TabView's tab bar on pushed levels (mirrors
/// Apple's iMessage / Music / Mail pattern). Tab root level keeps the
/// bar visible so the user can still switch tabs from there.
///
/// `.toolbar(.hidden, for: .tabBar)` hides the bar visually but doesn't
/// release its reserved safe-area space — that's a documented SwiftUI
/// behavior. We also `.ignoresSafeArea(.container, edges: .bottom)` so
/// the screen content actually fills the released vertical space.
/// Device-level safe areas (home indicator) are NOT in the `.container`
/// region, so they stay intact and our content still pins above the
/// home indicator correctly.
private struct HideTabBarOnPushModifier: ViewModifier {
    let isRoot: Bool

    func body(content: Content) -> some View {
        if isRoot {
            content
        } else {
            content
                .toolbar(.hidden, for: .tabBar)
                .ignoresSafeArea(.container, edges: .bottom)
        }
    }
}

/// Conditionally applies `.toolbarBackground` only when the layout
/// supplied an explicit color. Applying `.toolbarBackground(.clear, ...)`
/// to a default-styled bar disables iOS 26 Liquid Glass, which manifests
/// as the bar visibly going white during tab transitions and
/// republishes. Skipping the modifier entirely lets the system keep its
/// adaptive material.
private struct ExplicitBarBackgroundModifier: ViewModifier {
    let argb: Int
    let placement: ToolbarPlacement

    func body(content: Content) -> some View {
        if argb != 0 {
            content
                .toolbarBackground(Color(argb: argb), for: placement)
                .toolbarBackground(.visible, for: placement)
        } else {
            content
        }
    }
}

/// Encapsulates the iOS 26-only `.tabViewBottomAccessory(...)` and
/// `.tabBarMinimizeBehavior(...)` modifiers. Wrapping them in a single
/// `ViewModifier` keeps the `#available` branch confined to one place
/// and gives the renderer a no-op fallback on iOS 18-25.
private struct TabBarLiquidGlassModifier: ViewModifier {
    let accessory: NativeUINode?
    let minimizeOnScroll: Bool

    func body(content: Content) -> some View {
        if #available(iOS 26.0, *) {
            // `.tabViewBottomAccessory` always renders the slot — even when
            // its closure is empty — so an empty pill appears above the tab
            // bar. Only attach the modifier when the layout actually
            // supplied an accessory; otherwise just apply the minimize
            // behavior on its own.
            if let inner = accessory?.children.first {
                content
                    .tabBarMinimizeBehavior(minimizeOnScroll ? .onScrollDown : .never)
                    .tabViewBottomAccessory {
                        NodeView(node: inner)
                    }
            } else {
                content
                    .tabBarMinimizeBehavior(minimizeOnScroll ? .onScrollDown : .never)
            }
        } else {
            content
        }
    }
}

File: resources/androidstudio/app/src/main/java/com/nativephp/mobile/ui/nativerender/NativeRootTabsRenderer.kt
package com.nativephp.mobile.ui.nativerender

import androidx.activity.compose.BackHandler
import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.togetherWith
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.unit.dp
import androidx.compose.material3.Badge
import androidx.compose.material3.BadgedBox
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.NavigationBar
import androidx.compose.material3.NavigationBarDefaults
import androidx.compose.material3.NavigationBarItem
import androidx.compose.material3.NavigationBarItemDefaults
import androidx.compose.material3.Surface
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import com.nativephp.mobile.ui.MaterialIcon

/**
 * Compose port of iOS's `NativeRootTabsRenderer`. Renders the
 * `native_root_tabs` sentinel via Material 3 `Scaffold` +
 * `NavigationBar`, with an optional inner `TopAppBar` when the layout
 * supplies both bars.
 *
 * Search role and iOS 26-only modifiers (`tabViewBottomAccessory`,
 * `tabBarMinimizeBehavior`) have no exact Android equivalent; the
 * accessory renders inline above the NavigationBar, and search-flagged
 * tabs render as regular tabs.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun NativeRootTabsRenderer(node: NativeUINode, modifier: Modifier = Modifier) {
    val tabs = node.children.filter { it.type == "bottom_nav_item" }
    val accessory = node.children.firstOrNull { it.type == "tab_accessory" }
    // NavBar actions folded onto the tabs sentinel by `wrapWithNativeChrome`
    // when both bars are present — render trailing on the top bar.
    val actions = node.children.filter { it.type == "top_bar_action" }
    val screenContent = node.children.firstOrNull {
        it.type != "bottom_nav_item"
            && it.type != "top_bar_action"
            && it.type != "tab_accessory"
    }

    // Activeness flows from `BottomNavItem.active` (TabBar::highlight() set it).
    val activeTabIdx = tabs.indexOfFirst { it.props.getBool("active") }.coerceAtLeast(0)

    // Local selection mirrors PHP's active flag, but also responds to taps so
    // the bar UI updates instantly while we wait for PHP to republish.
    var selection by remember { mutableIntStateOf(activeTabIdx) }
    LaunchedEffect(activeTabIdx) {
        if (selection != activeTabIdx) selection = activeTabIdx
    }

    val activeColorArgb = node.props.getColor("active_color", 0)
    val bgArgb = node.props.getColor("background_color", 0)
    val textArgb = node.props.getColor("text_color", 0)

    // Per-screen explicit signal from PHP (`$hidesTabBar` shortcut or
    // `tabBarOptions()->hidden()` builder, folded into the chrome
    // sentinel by `wrapWithNativeChrome`). Replaces the earlier
    // URI-match / `nav_back` heuristics — those couldn't distinguish a
    // tab root with a back chevron from a pushed-detail screen.
    val hideTabBar = node.props.getBool("hide_tab_bar")

    // Active screen URI — used as part of the inner AnimatedContent's
    // key so within-tab navigation (chats list → chat detail) animates
    // even when the tab index doesn't change.
    val currentUri = node.props.getString("current_uri", "")

    // Folded NavBar config — present when the layout supplied both bars.
    val navBack = node.props.getBool("nav_back")
    val navTitle = node.props.getString("nav_title", "")
    val navSubtitle = node.props.getString("nav_subtitle", "")
    val navBgArgb = node.props.getColor("nav_background_color", 0)
    val navTextArgb = node.props.getColor("nav_text_color", 0)
    val hasNavBar = navBack || navTitle.isNotEmpty()

    // System back: defer to PHP (the tabs root has nowhere to pop within Compose).
    BackHandler(enabled = true) {
        NativeElementBridge.sendSystemBackEvent()
    }

    Scaffold(
        topBar = {
            if (hasNavBar) {
                TopAppBar(
                    title = {
                        if (navSubtitle.isNotEmpty()) {
                            Column {
                                Text(navTitle, fontWeight = FontWeight.SemiBold, style = MaterialTheme.typography.titleMedium)
                                Text(navSubtitle, style = MaterialTheme.typography.labelSmall)
                            }
                        } else {
                            Text(navTitle, fontWeight = FontWeight.SemiBold)
                        }
                    },
                    navigationIcon = {
                        if (navBack) {
                            IconButton(onClick = { NativeElementBridge.sendSystemBackEvent() }) {
                                Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back")
                            }
                        }
                    },
                    actions = {
                        actions.forEach { action -> TopBarActionView(action) }
                    },
                    colors = if (navBgArgb != 0) {
                        val bg = argbToComposeColor(navBgArgb)
                        val fg = if (navTextArgb != 0) argbToComposeColor(navTextArgb) else Color.White
                        TopAppBarDefaults.topAppBarColors(
                            containerColor = bg,
                            titleContentColor = fg,
                            navigationIconContentColor = fg,
                            actionIconContentColor = fg
                        )
                    } else {
                        TopAppBarDefaults.topAppBarColors()
                    }
                )
            }
        },
        bottomBar = bottomBar@{
            // Mirror iOS's `HideTabBarOnPushModifier`: pushed-detail
            // screens hide the tab strip entirely (iMessage / Music /
            // Mail pattern). Driven by the `hide_tab_bar` wire prop set
            // by the screen via `$hidesTabBar` or
            // `tabBarOptions()->hidden()`.
            if (hideTabBar) return@bottomBar

            Column {
                // Persistent accessory pinned above the NavigationBar
                // (Apple Music / Spotify / YouTube Music MiniPlayer
                // pattern). Compose / M3 has no native equivalent of
                // iOS 26's `.safeAreaBar(.bottom)` Liquid Glass capsule
                // — there's no built-in floating-bar primitive that
                // sits ABOVE the navigation bar, and Modifier.blur on
                // Android blurs the surface itself rather than the
                // content beneath it (so a true "backdrop blur" glass
                // material isn't directly available).
                //
                // Best-effort idiomatic replacement: a tonal Surface at
                // `surfaceContainerHigh` (one tonal step above the
                // NavigationBar's container so the two read as
                // separate layers) with rounded top corners + a small
                // shadow elevation. Matches what real-world Android
                // music apps ship.
                accessory?.children?.firstOrNull()?.let { acc ->
                    Surface(
                        color = MaterialTheme.colorScheme.surfaceContainerHigh,
                        shape = RoundedCornerShape(topStart = 16.dp, topEnd = 16.dp),
                        tonalElevation = 3.dp,
                        shadowElevation = 4.dp,
                        modifier = Modifier.fillMaxWidth(),
                    ) {
                        NodeView(node = acc)
                    }
                }

                NavigationBar(
                    containerColor = if (bgArgb != 0)
                        argbToComposeColor(bgArgb)
                    else
                        NavigationBarDefaults.containerColor
                ) {
                    tabs.forEachIndexed { idx, tab ->
                        val label = tab.props.getString("label", "")
                        val icon = tab.props.getString("icon", "circle")
                        val badge = tab.props.getString("badge", "")
                        val news = tab.props.getBool("news")

                        NavigationBarItem(
                            selected = idx == selection,
                            onClick = {
                                // Local selection updates instantly so the
                                // ripple / selection indicator responds; the
                                // press handler fires the PHP-side
                                // BottomNavItem-auto-wired `replace` nav (or
                                // a Tab::press() override on action tabs).
                                if (idx != activeTabIdx) {
                                    selection = idx
                                    if (tab.onPress != 0) {
                                        NativeElementBridge.sendPressEvent(tab.onPress, tab.id)
                                    }
                                }
                            },
                            icon = {
                                if (badge.isNotEmpty() || news) {
                                    BadgedBox(badge = {
                                        Badge {
                                            if (badge.isNotEmpty()) Text(badge)
                                        }
                                    }) {
                                        MaterialIcon(name = icon, contentDescription = label)
                                    }
                                } else {
                                    MaterialIcon(name = icon, contentDescription = label)
                                }
                            },
                            label = { Text(label) },
                            colors = if (activeColorArgb != 0) {
                                val active = argbToComposeColor(activeColorArgb)
                                NavigationBarItemDefaults.colors(
                                    selectedIconColor = active,
                                    selectedTextColor = active,
                                    indicatorColor = active.copy(alpha = 0.16f)
                                )
                            } else {
                                NavigationBarItemDefaults.colors()
                            }
                        )
                    }
                }
            }
        },
        modifier = modifier.fillMaxSize()
    ) { padding ->
        // Animate two distinct kinds of swap inside the tabs scaffold:
        //
        //   1. Tab switch (active tab index changes) — keep the soft
        //      cross-fade. PHP signals these via `BottomNavItem` with
        //      transition='none' because the renderer is supposed to
        //      handle the visual; honor that contract here.
        //
        //   2. Within-tab push / pop (active tab same, URI changed) —
        //      use the PHP-signaled transition (slide_from_right by
        //      default for navigate, slide_from_left for back). iOS gets
        //      this for free from per-tab NavigationStack; Android lacks
        //      that, so we drive it explicitly off the same
        //      `pendingTransition` that the root NativeUIContent reads.
        //      The bridge suppresses the *root* screenKey bump for
        //      native-chrome continuation publishes (NativeElementBridge
        //      ~line 261) so the outer animation stays out of our way;
        //      this inner one fills the gap.
        val pendingTransition by NativeUIBridge.pendingTransition
        val targetKey = "$activeTabIdx|$currentUri"

        AnimatedContent(
            targetState = targetKey,
            transitionSpec = {
                val initialIdx = initialState.substringBefore('|')
                val targetIdx = targetState.substringBefore('|')
                if (initialIdx != targetIdx) {
                    fadeIn(tween(180)) togetherWith fadeOut(tween(180))
                } else {
                    transitionFor(pendingTransition)
                }
            },
            label = "tab-content",
            modifier = Modifier.fillMaxSize().padding(padding)
        ) { _ ->
            if (screenContent != null) {
                NodeView(node = screenContent)
            } else {
                Box(modifier = Modifier.fillMaxSize())
            }
        }
    }
}
